### PR TITLE
Ouranos recipe for cookiecutter

### PR DESCRIPTION
**Pull request by Zeitsperre**
This PR is open just to advertise and discuss changes on this branch. I've added several changes inspired from xclim and miranda, but nothing exceptionally fancy, in my opinion.

There are definitely one or two changes that I might try pushing upstream (like typo fixes), but those will be in another branch before I try committing them to the parent project.

Are we on the right track here? What do we want to see? 
Related discussions:
https://github.com/Ouranosinc/xclim/issues/509
https://github.com/bird-house/cookiecutter-birdhouse/issues/99

### CODE REVIEW

**Comment by Zeitsperre** [comment]
It's in the dev recipe. If we want I can separate out the black configuration so that you are asked whether you want to use black at the start. The challenge would be adding it later, as the flake8 overrides would need to be added manually.

I also added black to the makefile `lint` recipe, but again, I can separate it from there as well.

**Comment by huard** [comment]
Nope, keep it as is, works for me.

**Comment by Zeitsperre** [comment]
@tlvu Should I be leveraging the makefile in Travis for the environment? The windows recipe is very specific to Travis-CI's vm structure, so I wouldn't move that there but should *nix-specific actions in the makefile be called from `.travis.yml`?

Should we be running checks against all the makefile commands?

File: {{cookiecutter.project_slug}}/.travis.yml [hunk 1]
```diff
@@ -1,11 +1,59 @@
...
+    - python: 3.8
+  allow_failures:
+    - env: TOXENV=black
+    - env: TOXENV=py38-macOS
+    - env:
+      - TOXENV=py38-Windows
+      - DESIRED_PYTHON=3.8
+      - MINICONDA_PATH=$(cygpath --windows /c/miniconda)
+
+before_install: printenv
```
Comment (tlvu) [inline {{cookiecutter.project_slug}}/.travis.yml:56]: nice !  Setting up tox cross-platform for Windows, Max as well, I think uptream will like this too.

File: {{cookiecutter.project_slug}}/Makefile [hunk 2]
```diff
@@ -49,6 +49,7 @@ clean-test: ## remove test and coverage artifacts
 
 lint: ## check style with flake8
 	flake8 {{ cookiecutter.project_slug }} tests
+	black --check --target-version py36 {{ cookiecutter.project_slug }} tests
```
Comment (tlvu) [inline {{cookiecutter.project_slug}}/Makefile:52]: this black line simply warn or will re-format the code?

File: {{cookiecutter.project_slug}}/setup.cfg [hunk 3]
```diff
@@ -11,18 +11,35 @@ replace = version='{new_version}'
 search = __version__ = '{current_version}'
 replace = __version__ = '{new_version}'
 
-[bdist_wheel]
-universal = 1
```
Comment (tlvu) [inline {{cookiecutter.project_slug}}/setup.cfg:15]: all changes in this `setup.cfg` is probably fairly specific to us, not applicable to upstream.

File: {{cookiecutter.project_slug}}/setup.py [hunk 4]
```diff
@@ -14,7 +14,11 @@
 
 setup_requirements = [{%- if cookiecutter.use_pytest == 'y' %}'pytest-runner',{%- endif %} ]
 
-test_requirements = [{%- if cookiecutter.use_pytest == 'y' %}'pytest>=3',{%- endif %} ]
+test_requirements = [{%- if cookiecutter.use_pytest == 'y' %}'pytest>=3',{%- endif %}]
```
Comment (tlvu) [inline {{cookiecutter.project_slug}}/setup.py:17]: all changes in this `setup.py` looks pretty generic, probable upstream would want it.

File: {{cookiecutter.project_slug}}/tox.ini [hunk 5]
```diff
@@ -1,30 +1,41 @@
...
-basepython = python
-deps = flake8
-commands = flake8 {{ cookiecutter.project_slug }} tests
+[testenv:black]
+skip_install = True
+deps =
+    flake8
+    black
+commands =
+    flake8 {{ cookiecutter.project_slug }} tests
```
Comment (tlvu) [inline {{cookiecutter.project_slug}}/tox.ini:17]: why not `make lint` in there to avoid repeating again the same commands in the Makefile?

File: {{cookiecutter.project_slug}}/tox.ini [hunk 6]
```diff
@@ -1,30 +1,41 @@
 [tox]
-envlist = py35, py36, py37, py38, flake8
+envlist = py{36,37,38}, black{%- if cookiecutter.make_docs ==  'y' %}, docs{%- endif %}, py38-{macOS,Windows}
```
Comment (tlvu) [inline {{cookiecutter.project_slug}}/tox.ini:2]: I think this ` {{cookiecutter.project_slug}}/tox.ini` can probably be submitted upstream as well.

**Comment by tlvu** [comment]
> @tlvu Should I be leveraging the makefile in Travis for the environment? The windows recipe is very specific to Travis-CI's vm structure, so I wouldn't move that there but should *nix-specific actions in the makefile be called from `.travis.yml`?
> 

You read my mind.

I think we should try our best to directly use the makefile in Travis-CI.  Otherwise, errors in Makefile will not be caught and dev workflow is essentially broken.  Basically one stone 2 birds, Travis-CI tests the product and the Makefile that builds the product at the same time.

Indeed, running make on Windows is trickier so just do what we can.

> Should we be running checks against all the makefile commands?

Yes, again only to the limit of reasonable possible, do not try 100% coverage of the Makefile.

That's why I added make commands to run docs and notebooks tests to Travis-CI in our cookiecutter-birdhouse.

All the major useful make commands should be continuously tested to ensure they are still working against continuous new dependencies update.  This allows us to not pin the versions in `requirements_dev.txt` and `requirements_docs.txt`.

File: tox.ini [hunk 7]
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py35, py36, py37,py38 pypy, docs
+envlist = py35, py36, py37, py38, pypy, docs
```
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply tox.ini:2]: It already is, but the PR has been open for many months now. At the very least, once that's merged, nothing will change here.

File: .travis.yml [hunk 8]
```diff
@@ -17,7 +21,7 @@ script: tox
 deploy:
   provider: pypi
   distributions: sdist bdist_wheel
-  user: audreyr
+  username: audreyr
```
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply .travis.yml:24]: It raises a warning as the API for Travis CI has changed in the past year. I'll check and see if this isn't already in the PR queue upstream.

File: {{cookiecutter.project_slug}}/Makefile (hunk 2 above)
Comment (Zeitsperre) [reply {{cookiecutter.project_slug}}/Makefile:52]: This is simply a warning. Linting code shouldn't reformat it.

File: {{cookiecutter.project_slug}}/setup.cfg (hunk 3 above)
Comment (Zeitsperre) [reply {{cookiecutter.project_slug}}/setup.cfg:15]: The `bdist_wheel` flag is no longer relevant. It's only meant for packaging python2/3-compatible packages. https://packaging.python.org/guides/distributing-packages-using-setuptools/#universal-wheels

File: {{cookiecutter.project_slug}}/tox.ini (hunk 5 above)
Comment (Zeitsperre) [reply {{cookiecutter.project_slug}}/tox.ini:17]: Oh dang. Not a bad idea!

File: {{cookiecutter.project_slug}}/setup.cfg (hunk 3 above)
Comment (tlvu) [reply {{cookiecutter.project_slug}}/setup.cfg:15]: Yes sorry unclear my bad, this universal wheel is okay, I meant the rest of the file, like all the specific flakes8 ignore rules, that's more specific to us.  I should have chosen a different diff hunk to put my comment.  There is no file wide comment so I had to attach to one of the diff hunk.

Comment (Zeitsperre) [reply {{cookiecutter.project_slug}}/setup.cfg:15]: Ah, right. The flake8 overrides are needed specifically for black and flake8 to play nicely. If we were to push black into upstream, _most_ of these changes would be needed. Worth considering.

File: {{cookiecutter.project_slug}}/setup.py (hunk 4 above)
Comment (Zeitsperre) [reply {{cookiecutter.project_slug}}/setup.py:17]: Maybe!

File: {{cookiecutter.project_slug}}/setup.cfg (hunk 3 above)
Comment (tlvu) [reply {{cookiecutter.project_slug}}/setup.cfg:15]: Oh did not realize these flake8 ignore are for black.  Then yes if we push black upstream it should go together.  Same as for all the changes in tox.ini, I think most is for black.

So we should have a few separate PR for upstream, one for the typo, deprecation stuff, very high chance of acceptance, 1 for the Windows and Mac Travis-CI support, good chance of acceptance, and finally all things related to black, probably make it a toggle so higher acceptance chance.

In general small, easy to wrap our head around and focused PR have higher chance of acceptance than big, hard to follow which diff hunks are needed for which intend PR.
